graphql/schemabuilder/pagination.go

```diff
@@ -0,0 +1,530 @@
+func EdgesToReturn(allEdges []Edge, before *string, after *string, first *int64, last *int64) ([]Edge, bool, bool, error) {
+
+	edges, elemsAfter, elemsBefore := applyCursorsToAllEdges(allEdges, before, after)
+
+	prevPage := false
+	nextPage := false
+
+	if first != nil {
+		if *first < 0 {
+			return nil, nextPage, prevPage, graphql.NewClientError("first should be a non-negative integer")
+		}
+		if len(edges) > int(*first) {
+			edges = edges[:int(*first)]
+			nextPage = true
+		}
+	}
+	if before != nil {
+		if elemsAfter {
+			nextPage = true
+		}
+	}
+
+	if last != nil {
+		if *last < 0 {
+			return nil, nextPage, prevPage, graphql.NewClientError("last should be a non-negative integer")
+		}
+		if len(edges) > int(*last) {
+			edges = edges[len(edges)-int(*last):]
+			prevPage = true
+		}
+	}
+	if after != nil {
+		if elemsBefore {
+			prevPage = true
+		}
+	}
+
+	return edges, nextPage, prevPage, nil
+
+}
+
+// getCursorIndex returns the index corresponding to the cursor in the slice.
+func getCursorIndex(edges []Edge, cursor string) int {
+
+	for i, val := range edges {
+		if val.Cursor == cursor {
+			return i
+		}
+	}
+	return -1
+}
+
+// applyCursorsToAllEdges returns the slice of edges after applying the after and before arguments.
+// It also implements part of the hasNextPage and hasPrevPage algorithm by returning if there are
+// elements after or before the arguments.
+func applyCursorsToAllEdges(allEdges []Edge, before *string, after *string) ([]Edge, bool, bool) {
+
+	edges := allEdges
+
+	elemsAfter := false
+	elemsBefore := false
+
+	if after != nil {
+		i := getCursorIndex(edges, *after)
+		if i != -1 {
+			edges = edges[i+1:]
+			if i != 0 {
+				elemsBefore = true
+			}
+		}
+
+	}
+	if before != nil {
+		i := getCursorIndex(edges, *before)
+		if i != -1 {
+			edges = edges[:i]
+			if i != len(allEdges)-1 {
+				elemsAfter = true
+			}
+		}
+
+	}
+
+	return edges, elemsAfter, elemsBefore
+
+}
+
+// getConnection applies the ConnectionArgs to nodes and returns the result in a wrapped Connection
+// type.
+func getConnection(key string, nodes []interface{}, args ConnectionArgs) (interface{}, error) {
+
+	var edges []Edge
+	for _, val := range nodes {
+		// Get the value of the key field and then b64 encode it for the cursor.
+		keyValue := reflect.ValueOf(val)
+		if keyValue.Kind() == reflect.Ptr {
+			keyValue = keyValue.Elem()
+		}
+		keyString := []byte(fmt.Sprintf("%v", keyValue.FieldByName(key).Interface()))
+		edges = append(edges, Edge{Node: val, Cursor: base64.StdEncoding.EncodeToString(keyString)})
+	}
+
+	var err error
+	var nextPage, prevPage bool
+	edges, nextPage, prevPage, err = EdgesToReturn(edges, args.Before, args.After, args.First, args.Last)
+
+	endCursor := ""
+	if len(edges) > 0 {
+		endCursor = edges[len(edges)-1].Cursor
+	}
+	startCursor := ""
+	if len(edges) > 0 {
+		startCursor = edges[0].Cursor
+	}
+
+	if err != nil {
```

Comment:
move this up before the startCursor/endCursor stuff, so it's clear which error you're checking

Reply:
done